[llvm][CodeGen] Intrinsic `llvm.powi.*` code gen for vector arguments

ylzsx opened the pull request:
Scalarize vector FPOWI instead of promoting the type. This allows the scalar FPOWIs to be visited and converted to libcalls before promoting the type.

FIXME: This should be done in LegalizeVectorOps/LegalizeDAG, but call lowering needs the unpromoted EVT.

Without this patch, in some backends, such as RISCV64 and LoongArch64, the i32 type is illegal and will be promoted. This causes exponent type check to fail when ISD::FPOWI node generates a libcall.

Fix https://github.com/llvm/llvm-project/issues/118079

ylzsx committed:
[llvm][CodeGen] Intrinsic `llvm.powi.*` code gen for vector arguments

In some backends, the i32 type is illegal and will be promoted. This
causes exponent type check to fail when ISD::FOWI node generates a libcall.

ylzsx committed:
Modify test file name. NFC

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -4648,6 +4648,24 @@ void SelectionDAGLegalize::ConvertNodeToLibcall(SDNode *Node) {
+      // generated.
+      SDNode *ExponentNode = Node->getOperand(1 + Offset).getNode();
+      unsigned LibIntSize = DAG.getLibInfo().getIntSize();
+      if (ExponentNode->getOpcode() == ISD::SIGN_EXTEND_INREG ||
+          ExponentNode->getOpcode() == ISD::AssertSext ||
+          ExponentNode->getOpcode() == ISD::AssertZext) {
+        EVT InnerType = cast<VTSDNode>(ExponentNode->getOperand(1))->getVT();
+        ExponentHasSizeOfInt = LibIntSize == InnerType.getSizeInBits();
+      } else if (ISD::isExtOpcode(ExponentNode->getOpcode())) {
+        ExponentHasSizeOfInt =
+            LibIntSize ==
+            ExponentNode->getOperand(0).getValueType().getSizeInBits();
+      }
```
This code should not be trying to look through extensions (nor SIGN_EXTEND_INREG, or the Asserts*. These aren't really extensions).

I'd expect this to just insert the sext to match the libcall integer type

topperc commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -4648,6 +4648,24 @@ void SelectionDAGLegalize::ConvertNodeToLibcall(SDNode *Node) {
         DAG.getLibInfo().getIntSize() ==
         Node->getOperand(1 + Offset).getValueType().getSizeInBits();
+    if (!ExponentHasSizeOfInt) {
+      // In some backends, such as RISCV64 and LoongArch64, the i32 type is
+      // illegal and is promoted by previous process. For such cases, the
+      // exponent actually matches with sizeof(int) and a libcall should be
```
Why would it be 8? getIntSize is in bits and `int` on RISCV64 is 32 bits.

arsenm replied:
I would assume getIntSize() would refer to the legalized parameter type for a libcall using int. Is there a second version for this?

topperc replied:
Not that I know of.

I don't think we should let the vector powi get thisfar . The promotion of a libcall argument is the responsibility of call lowering and the information in TargetLowering::MakeLibCallOptions. We shouldn't promote and then try to fix it.

topperc commented:
I kind of think this should be handled by unrolling the vector in `DAGTypeLegalizer::PromoteIntOp_ExpOp`.

ylzsx commented:
> I kind of think this should be handled by unrolling the vector in `DAGTypeLegalizer::PromoteIntOp_ExpOp`.
> 
> This should work
> 
> ```
> diff --git a/llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp b/llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp
> index 493abfde148c..e386fd510419 100644
> --- a/llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp
> +++ b/llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp
> @@ -2579,6 +2579,9 @@ SDValue DAGTypeLegalizer::PromoteIntOp_ExpOp(SDNode *N) {
>        N->getOpcode() == ISD::FPOWI || N->getOpcode() == ISD::STRICT_FPOWI;
>    unsigned OpOffset = IsStrict ? 1 : 0;
> 
> +  if (N->getValueType(0).isVector())
> +    return DAG.UnrollVectorOp(N);
> +
>    // The integer operand is the last operand in FPOWI (or FLDEXP) (so the result
>    // and floating point operand is already type legalized).
>    RTLIB::Libcall LC = IsPowI ? RTLIB::getPOWI(N->getValueType(0))
> ```

Thanks. Your approach is simpler and more reasonable. I will modify it to adopt this solution.

ylzsx committed:
Fixes for reviews.

For nodes such as `ISD::FPOWI`, `ISD::FLDEXP`, if the first operand is
a vector operand, since the corresponding library functions do not have
vector-type signatures, the vector will be unroll during the type
legalization, without promoting the second operand.

topperc commented on llvm/test/CodeGen/RISCV/rvv/fixed-vectors-fldexp.ll:
```diff
@@ -0,0 +1,1427 @@
+; RUN: llc -mtriple=riscv64 -mattr=+v,+f,+d -target-abi=lp64d -verify-machineinstrs < %s \
+; RUN:   | FileCheck %s --check-prefix=RV64
+
+define <1 x float> @ldexp_v1f32(<1 x float> %x, i32 %y) {
```
Add `nounwind` to these tests

topperc commented on llvm/test/CodeGen/RISCV/rvv/fixed-vectors-fpowi.ll:
```diff
@@ -0,0 +1,1427 @@
+; RUN: llc -mtriple=riscv64 -mattr=+v,+f,+d -target-abi=lp64d -verify-machineinstrs < %s \
+; RUN:   | FileCheck %s --check-prefix=RV64
+
+define <1 x float> @powi_v1f32(<1 x float> %x, i32 %y) {
```
Add `nounwind`

ylzsx replied:
Thanks, I will modify it.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -2572,6 +2572,9 @@ SDValue DAGTypeLegalizer::PromoteIntOp_FRAMERETURNADDR(SDNode *N) {
 }
 
 SDValue DAGTypeLegalizer::PromoteIntOp_ExpOp(SDNode *N) {
+  if (N->getValueType(0).isVector())
+    return DAG.UnrollVectorOp(N);
```
This probably shouldn't be the first thing tried. If a vector libcall happens to be available, that would be preferable. Can you move this down before the scalar call is introduced?

ylzsx replied:
Thanks, I will move it into this `if` statement to ensure that it won't promote the second operand. Do you think it's reasonable?
```
  if (LC == RTLIB::UNKNOWN_LIBCALL || !TLI.getLibcallName(LC)) {
    if (N->getValueType(0).isVector())
      return DAG.UnrollVectorOp(N);
    SmallVector<SDValue, 3> NewOps(N->ops());
    NewOps[1 + OpOffset] = SExtPromotedInteger(N->getOperand(1 + OpOffset));
    return SDValue(DAG.UpdateNodeOperands(N, NewOps), 0);
  }
```

ylzsx committed:
Fixes for review.

ylzsx committed:
code formatter

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -2585,6 +2585,8 @@ SDValue DAGTypeLegalizer::PromoteIntOp_ExpOp(SDNode *N) {
                              : RTLIB::getLDEXP(N->getValueType(0));
 
   if (LC == RTLIB::UNKNOWN_LIBCALL || !TLI.getLibcallName(LC)) {
+    if (N->getValueType(0).isVector())
+      return DAG.UnrollVectorOp(N);
```
This whole function is actually the wrong place to split the vector (see how there are no other UnrollVectorOps uses in DAGTypeLegalizer). The description also says your problem is when the libcall is used, so you'd want to change the other path? Does the libcall emission below need to directly handle the vector case?

ylzsx replied:
Sorry, I'm not entirely sure I understand what you mean. Could you please clarify? 
Are you referring to the approach used in my first commit https://github.com/llvm/llvm-project/pull/118242/commits/7f5a128d9c3f9468f7a4c6bd2c08dfa478f3bf5c ? If so, in that commit, you mentioned not to check for the `SIGN_EXTEND_INREG` node, but this node is generated in this `if `statement here.(`SExtPromotedInteger` will generate a `SIGN_EXTEND_INREG` node). 
```
  if (LC == RTLIB::UNKNOWN_LIBCALL || !TLI.getLibcallName(LC)) {
    SmallVector<SDValue, 3> NewOps(N->ops());
    NewOps[1 + OpOffset] = SExtPromotedInteger(N->getOperand(1 + OpOffset));
    return SDValue(DAG.UpdateNodeOperands(N, NewOps), 0);
  }
```

Other info:
1. RISCV64 and Loongarch64 will enter function `PromoteIntOp_ExpOp` and generate a `SIGN_EXTEND_INREG` node, so `ExponentHasSizeOfInt=false`(32 != 64) and report an error.
```
bool ExponentHasSizeOfInt =
        DAG.getLibInfo().getIntSize() ==
        Node->getOperand(1 + Offset).getValueType().getSizeInBits();
```
Additionally, AArch64 and X86 will get `ExponentHasSizeOfInt=true`, because in these backends, i32 is valid and was not promoted to `SIGN_EXTEND_INREG`(i64) previously. So, they generate right code sequences.

2. In the entire LLVM, only `fpowi` and `fldexp` will call the `PromoteIntOp_ExpOp` function.(Flow: `llvm::DAGTypeLegalizer::PromoteIntegerOperand` -> `llvm::DAGTypeLegalizer::PromoteIntOp_ExpOp`)
```
  case ISD::FPOWI:
  case ISD::STRICT_FPOWI:
  case ISD::FLDEXP:
  case ISD::STRICT_FLDEXP: Res = PromoteIntOp_ExpOp(N); break;
```

If not, could you provide more specific modification suggestions？

topperc replied:
> see how there are no other UnrollVectorOps uses in DAGTypeLegalizer

There are calls to `UnrollVectorOps` in several places in LegalizeVectorTypes.cpp

This comment where the libcall below is created seems relevant

```
  // We can't just promote the exponent type in FPOWI, since we want to lower
  // the node to a libcall and we if we promote to a type larger than
  // sizeof(int) the libcall might not be according to the targets ABI.
```

My suggestion to unroll here was so that we wouldn't promote past sizeof(int). If we wait until LegalizeDAG.cpp to unroll the operation, the damage to the integer type has already been done. For RISC-V its harmless because `signed int` is supposed to be passed sign extended to 64-bits according to the ABI.

Hypothetically, if powi took an `unsigned int` as an argument, then type legalization would use zero extend, but the RISC-V ABI wants `unsigned int` to be passed sign extended. So LegalizeDAG would need to insert a SIGN_EXTEND_INREG to fix. I guess it would need to use the getIntSize() and shouldSignExtendTypeInLibCall to know what it needs to do in that case.

If we don't unroll here I guess the best fix in LegalizeDAG would also be to use `getIntSize()` and `shouldSignExtendTypeInLibCall` and use computeNumSignBits to know if a SIGN_EXTEND_INREG needs to be inserted?

arsenm replied:
It makes more sense to me to handle this when emitting the call, where ABI constraints would naturally be handled. This can be sign extended here, and the call emission can truncate or sext_inreg as required

topperc replied:
> It makes more sense to me to handle this when emitting the call, where ABI constraints would naturally be handled. This can be sign extended here, and the call emission can truncate or sext_inreg as required

The POWI code in LegalizeDAG calls `SelectionDAGLegalize::ExpandFPLibCall` which  will call `TargetLowering::makeLibCall` using the promoted type. `makeLibCall` calls `getTypeForEVT` which will return `i64` due to the promotion. That's what will be used by calling lowering, but that's the wrong type for it to do the right thing. We need to get an i32 Type* into call lowering, but we no longer have it. We'll need to call getIntSize() to get the size and pass it along somehow. That requires refactoring several interfaces or adding new ones. Not sure if call lowering would also expect the SDValue for the argument to have i32 type.

My unrolling proposal avoided that by scalarizing it and letting the newly created scalar powi calls get converted to libcalls while we're still in type legalization. That's how we currently handle the scalar powi case. Are you also suggesting we should move the scalar handling to LegalizeDAG as well?

arsenm replied:
makeLibCall should really be looking at the signature of the underlying call, but RuntimeLibcalls currently does not record this information. TargetLibraryInfo does, which is separate for some reason. This keeps coming up as a problem, these really need to be merged in some way (cc @jhuber6).

Taking the type from the DAG node isn't strictly correct, it's just where we've ended up. This came up recently for the special case in ExpandFPLibCall to sign extend the integer argument for FLDEXP. 

Practically speaking, I don't think any targets will have a vector powi implementation (I at least don't see any in RuntimeLibcalls), so unrolling works out. I guess this could get a fixme and go ahead for now, but it's still a hack

arsenm reviewed (commented):
I just noticed you are using vector ldexp with a scalar i32 argument. I'm surprised this is considered valid IR, and works as well as it does. It wasn't my intention when I added the intrinsic to handle that case. I expected the number of elements of both arguments would match. Do you only have this issue when using the implicit splat behavior of the scalar operand? powi is weird because it only takes the scalar argument, unlike ldexp.

I suppose we could support the scalar second argument for ldexp, but I'm not sure we handle implicit splats like that in any other operation.

ylzsx committed:
Intrinsic `llvm.ldexp.*` should not be affected.

topperc commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -2585,6 +2585,8 @@ SDValue DAGTypeLegalizer::PromoteIntOp_ExpOp(SDNode *N) {
                              : RTLIB::getLDEXP(N->getValueType(0));
 
   if (LC == RTLIB::UNKNOWN_LIBCALL || !TLI.getLibcallName(LC)) {
+    if (IsPowI && N->getValueType(0).isVector())
```
Add a comment something like this

```
// Scalarize vector FPOWI instead of promoting the type. This allows the scalar FPOWIs to be visited and converted to libcalls before promoting the type.
FIXME: This should be done in LegalizeVectorOps/LegalizeDAG, but call lowering needs the unpromoted EVT.
```

ylzsx replied:
Thanks, I will add the comment.

ylzsx committed:
Add comment.

ylzsx commented:
ping

topperc reviewed (approved):
LGTM

SixWeining reviewed (approved):
LGTM
